### PR TITLE
Check CisStatusScan Fail field before acccesing

### PR DESCRIPTION
**Pull request by nickgerace**
Check if the CisStatusScan Fail field is initialized before accessing
it. If Fail is "nil", a panic will occur.

Issue: #29518

### CODE REVIEW

**Review by prachidamle — DISMISSED** [review]
LGTM

**Comment by nickgerace** [comment]
CC: @dramich @prachidamle...

I believe we may either want to...

1. keep the PR "as-is" (or, at most, refactor the existing logic) because the ```ClusterScanRule``` is checked before calling ```isAlertRuleMatching```.
2. move the ```alertRule.Spec.ClusterScanRule == nil``` check inside ```isAlertRuleMatching```.
3. put a new ```cs.Status.CisScanStatus == nil``` check in the outer function (```Sync```).

Here's the check for the ```ClusterScanRule```...
```go
if alertRule.Status.AlertState == "inactive" || alertRule.Spec.ClusterScanRule == nil {
	continue
}
```

The last option would reduce number of calls on the function call stack. We may not even need the ```isAlertRuleMatching``` in that case. Here's the logic for that function...
```go
if alertRule.Spec.ClusterScanRule.ScanRunType != cs.Spec.RunType {
	return false
}
if alertRule.Spec.ClusterScanRule.FailuresOnly && !(cs.Status.CisScanStatus.Fail > 0) {
	return false
}
return true
```

So then the final design could incorporate that logic inside ```Sync```.
```go
for _, alertRule := range clusterAlertRules {
    // insert isAlertRuleMatching logic here
}
```

Reference: https://github.com/rancher/rancher/blob/master/pkg/controllers/managementuser/alert/watcher/cluster_scan.go#L76

**Review by rmweir — DISMISSED** [review]
LGTM

**Comment by nickgerace** [comment]
Need to change this PR in a oder to check if CisScanStatus is not nil before calling the function in question.

**Comment by nickgerace** [comment]
Changed the PR after discussing with @prachidamle and @dramich. Cis sends alerts when the scan has completed and when the scan has failures. In both scenarios, the Status should be set, so I've added a skip at the top of the controller that triggers is Status is nil.

File: pkg/controllers/managementuser/alert/watcher/cluster_scan.go
Comment (dramich) [inline pkg/controllers/managementuser/alert/watcher/cluster_scan.go:53]: Let's add a comment on why we skip the object if `CisScanStatus` is nil

Comment (nickgerace) [reply pkg/controllers/managementuser/alert/watcher/cluster_scan.go:53]: Will do.

Comment (nickgerace) [reply pkg/controllers/managementuser/alert/watcher/cluster_scan.go:56]: Done.

**Review by dramich — DISMISSED** [review]

File: pkg/controllers/managementuser/alert/watcher/cluster_scan.go
Comment (rmweir) [inline pkg/controllers/managementuser/alert/watcher/cluster_scan.go:56]: Do we know if the object will be modified once it is finished or should we be reenqueueing the object?

Comment (rmweir) [inline pkg/controllers/managementuser/alert/watcher/cluster_scan.go:56]: nit: I'd probably just separate these checks since the comment only apply to one of them.

**Review by rmweir — APPROVED** [review]
LGTM, one nit

File: pkg/controllers/managementuser/alert/watcher/cluster_scan.go
Comment (nickgerace) [reply pkg/controllers/managementuser/alert/watcher/cluster_scan.go:56]: Hmmmmmmmm, I concur, Ricky. Will do.

Comment (nickgerace) [reply pkg/controllers/managementuser/alert/watcher/cluster_scan.go]: Not the cleanest thing ever, but I think the readability is more important here, and agree with you.

**Review by rmweir — APPROVED** [review]
LGTM

**Comment by nickgerace** [comment]
Merging after blue, green, and Dan approval and "go ahead" to merge. Changes have been cosmetic since Dan's approval.

**Pull request merged**
